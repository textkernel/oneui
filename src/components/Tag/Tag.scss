.Tag {
    display: flex;
    align-items: center;

    padding: var(--spacing-normal) var(--spacing-2x);
    border-radius: var(--border-radius);

    &--isSelected {
        box-shadow: 0 0 0 2px var(--color-primary);
    }

    &--clickable {
        cursor: pointer;
    }

    &__text{
        margin: 0;

        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__deleteButton {
        display: flex;
        align-items: center;

        min-width: 15px;
        min-height: 15px;

        border: 0;
        cursor: pointer;
        background-color: var(--transparent);

        padding: 0;
        margin-left: var(--spacing-normal);
    }
}
