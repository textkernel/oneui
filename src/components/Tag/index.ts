export { Tag, Props as TagProps } from './Tag';
