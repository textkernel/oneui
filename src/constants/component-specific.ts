/** Constants that are component specific but do not merit a separate file */

/** Autocomplete */
export const NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS = 5;

/** Button */
export const BUTTON_TYPES = ['submit', 'button'] as const;

export type ButtonType = typeof BUTTON_TYPES[number];

/** Input */
export const INPUT_TYPES = [
    'text',
    'password',
    'email',
    'date',
    'datetime-local',
    'month',
    'tel',
    'number',
    'url',
    'search',
    'time',
    'week',
    'color',
] as const;

export type InputType = typeof INPUT_TYPES[number];

/** List */
export const LIST_NAVIGATION_DIRECTIONS = {
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};

/** PopupBase, MultiActionButton */
export const POPUP_PLACEMENTS = [
    'auto',
    'auto-start',
    'auto-end',
    'top',
    'top-start',
    'top-end',
    'right',
    'right-start',
    'right-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'left',
    'left-start',
    'left-end',
] as const;

export type PopupPlacement = typeof POPUP_PLACEMENTS[number];

/** Gauge */
export const GAUGE_RADIUS = 100;
