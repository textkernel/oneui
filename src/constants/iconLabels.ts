export const CLOSE_ICON_LABEL = 'X';
export const UP_ARROW_LABEL = 'arrow up';
export const DOWN_ARROW_LABEL = 'arrow down';
