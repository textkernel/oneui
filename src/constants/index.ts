export * from './style-related';
export * from './component-specific';
export * from './keyboard';
export * from './characters';
