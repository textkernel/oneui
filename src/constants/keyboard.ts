export const ENTER_KEY = 'Enter';
export const BACKSPACE_KEY = 'Backspace';
export const ESCAPE_KEY = 'Escape';
export const TAB_KEY = 'Tab';
export const SPACE_KEY = ' ';
