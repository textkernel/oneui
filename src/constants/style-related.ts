export const VALIDATION_CONTEXTS = ['success', 'warning', 'danger'] as const;

export const CONTEXTS = ['info', ...VALIDATION_CONTEXTS] as const;

export const SIZES = ['small', 'normal', 'large'] as const;
export const HEADING_SIZES = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'] as const;
export const CHECKBOX_VIEWBOX = [0, 0, 10, 10] as const;
export const COLORS = [
    '#AFF5E2',
    '#D6F481',
    '#FFF16B',
    '#FFD571',
    '#FFC4A6',
    '#E5DAC9',
    '#ECCBFF',
    '#FFC5EB',
    '#FCC1CF',
    '#C4D7FC',
    '#ADDDFF',
    '#75DFFB',
];

export type ValidationContext = (typeof VALIDATION_CONTEXTS)[number];
export type Context = (typeof CONTEXTS)[number];
export type Size = (typeof SIZES)[number];
