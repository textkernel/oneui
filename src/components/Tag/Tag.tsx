import * as React from 'react';
import { MdClose } from 'react-icons/md';
import { bem } from '../../utils';
import { Text } from '../Text';
import { ENTER_KEY } from '../../constants';
import styles from './Tag.scss';

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
    /** Value that helps us to understand if the component is currently selected or not. */
    isSelected?: boolean;
    /** Color is assigned to a background color of a wrapper */
    bgColor?: string;
    /** Max-width of a component */
    maxWidth?: string;
    /** Size of the text */
    size?: 'small' | 'normal' | 'large';
    /** Callback, that is fired when a user clicks on a delete icon */
    onDelete?: (e: React.KeyboardEvent | React.MouseEvent) => void;
    /** Callback, that is fired when a user clicks on an element */
    onClick?: (e: React.KeyboardEvent | React.MouseEvent) => void;
    /** A css class to be applied to the content (child) */
    contentClassName?: string;
    /** A css style to be applied to the content (child) */
    contentStyle?: React.CSSProperties;
}

const { block, elem } = bem('Tag', styles);

export const Tag = React.forwardRef<HTMLDivElement, Props>((props, ref) => {
    const {
        children,
        bgColor,
        maxWidth,
        size,
        onDelete,
        onClick,
        isSelected,
        contentClassName,
        contentStyle,
        ...rest
    } = props;

    // Generate title for children that are plain text (without tags)
    // If there is something different from the string (JSX) - children will be of object type.
    const areChildrenString = typeof children === 'string';

    const handleDeleteClick = (e: React.MouseEvent<HTMLElement>) => {
        e.stopPropagation();

        onDelete?.(e);
    };

    const handleTagKeyPress = (e: React.KeyboardEvent) => {
        e.stopPropagation();

        if (e.key === ENTER_KEY) {
            onClick?.(e);
        }
    };

    const handleDeleteButtonKeyPress = (e: React.KeyboardEvent) => {
        e.stopPropagation();
        e.preventDefault();

        if (e.key === ENTER_KEY) {
            onDelete?.(e);
        }
    };

    return (
        <div
            {...rest}
            ref={ref}
            {...block({ ...props, clickable: !!onClick })}
            {...(onClick && {
                onClick,
                tabIndex: 0,
                role: 'button',
                onKeyPress: handleTagKeyPress,
            })}
            style={{
                backgroundColor: bgColor,
                maxWidth,
            }}
        >
            <Text
                size={size}
                {...(areChildrenString && { title: children })}
                {...elem('text', { elemClassName: contentClassName })}
                style={contentStyle}
            >
                {children}
            </Text>
            {onDelete && (
                <button
                    onClick={handleDeleteClick}
                    onKeyDown={handleDeleteButtonKeyPress}
                    type="button"
                    {...elem('deleteButton')}
                >
                    <MdClose size="15px" />
                </button>
            )}
        </div>
    );
});

Tag.displayName = 'Tag';

Tag.defaultProps = {
    isSelected: false,
    bgColor: 'var(--color-background)',
    maxWidth: 'fit-content',
    size: 'normal',
    onDelete: undefined,
    onClick: undefined,
};
