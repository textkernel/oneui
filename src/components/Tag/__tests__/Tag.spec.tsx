import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { ENTER_KEY } from '../../../constants';
import { Tag } from '../Tag';

describe('<Tag> component', () => {
    let view: RenderResult;
    const onTagClick = jest.fn();
    const onDeleteClick = jest.fn();

    const text = 'Tag text';
    const bgColor = '#ccc';
    const maxWidth = '30px';
    const textSize = 'large';
    const closeLabel = 'close label';

    it('should render correctly with minimum amount of props', () => {
        view = render(<Tag>{text}</Tag>);

        expect(screen.getByText(text)).toBeInTheDocument();
        expect(view.container).toMatchSnapshot();
    });

    it('should render correctly with full list of props', () => {
        view = render(
            <Tag
                isSelected
                bgColor={bgColor}
                maxWidth={maxWidth}
                size={textSize}
                onClick={onTagClick}
                onDelete={onDeleteClick}
                contentClassName="my-class"
                contentStyle={{ color: 'red' }}
            >
                {text}
            </Tag>
        );

        const button = screen.getByRole('button', { name: text });

        expect(view.container).toMatchSnapshot();
        expect(button).toBeInTheDocument();
        expect(button).toHaveAttribute(
            'style',
            'background-color: rgb(204, 204, 204); max-width: 30px;'
        );
    });

    it('should invoke callback when onClick event is called', async () => {
        const user = userEvent.setup();
        view = render(<Tag onClick={onTagClick}>{text}</Tag>);

        await user.click(screen.getByRole('button'));

        expect(onTagClick).toHaveBeenCalledTimes(1);
    });

    it('should invoke onClick when user interacts with tag by pressing Enter', async () => {
        const user = userEvent.setup();
        view = render(<Tag onClick={onTagClick}>{text}</Tag>);

        const button = screen.getByRole('button');
        await button.focus();
        await user.keyboard(`[${ENTER_KEY}]`);

        expect(onTagClick).toHaveBeenCalledTimes(1);
    });

    it('should invoke callback when onDelete event is called', async () => {
        const user = userEvent.setup();
        view = render(
            <Tag onClick={onTagClick} onDelete={onDeleteClick}>
                {text}
            </Tag>
        );

        await user.click(screen.getByRole('button', { name: '' }));

        expect(onTagClick).toHaveBeenCalledTimes(0);
        expect(onDeleteClick).toHaveBeenCalledTimes(1);
    });

    it('should invoke onDelete when user interacts with delete button by pressing Enter', async () => {
        const user = userEvent.setup();
        view = render(
            <Tag onClick={onTagClick} onDelete={onDeleteClick} closeLabel={closeLabel}>
                {text}
            </Tag>
        );
        const button = screen.getByRole('button', { name: closeLabel });

        await button.focus();
        await user.keyboard(`[${ENTER_KEY}]`);

        expect(onTagClick).toHaveBeenCalledTimes(0);
        expect(onDeleteClick).toHaveBeenCalledTimes(1);
    });
});
