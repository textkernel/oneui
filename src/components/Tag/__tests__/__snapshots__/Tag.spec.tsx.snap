// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag> component should render correctly with full list of props 1`] = `
<div>
  <div
    class="Tag Tag--isSelected Tag--clickable"
    role="button"
    style="background-color: rgb(204, 204, 204); max-width: 30px;"
    tabindex="0"
  >
    <p
      class="Text--size_large Tag__text my-class"
      style="color: red;"
      title="Tag text"
    >
      Tag text
    </p>
    <button
      class="Tag__deleteButton"
      type="button"
    >
      <svg
        fill="currentColor"
        height="15px"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="15px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`<Tag> component should render correctly with minimum amount of props 1`] = `
<div>
  <div
    class="Tag"
    style="max-width: fit-content;"
  >
    <p
      class="Tag__text"
      title="Tag text"
    >
      Tag text
    </p>
  </div>
</div>
`;
